# Исходные данные
Датасет: https://www.openml.org/d/44

Предметная область: спам в электронной рассылке.
Задача: определить, является ли электронное письмо спамом.
Записей: 4601, атрибутов: 57.

## Атрибуты
1.	Частота использования строки «make» (вещественный тип, [0,100])
2.	Частота использования строки «address» (вещественный тип, [0,100])
3.	Частота использования строки «all» (вещественный тип, [0,100])
4.	Частота использования строки «3d» (вещественный тип, [0,100])
5.	Частота использования строки «our» (вещественный тип, [0,100])
6.	Частота использования строки «over» (вещественный тип, [0,100])
7.	Частота использования строки «remove» (вещественный тип, [0,100])
8.	Частота использования строки «internet» (вещественный тип, [0,100])
9.	Частота использования строки «order» (вещественный тип, [0,100])
10.	Частота использования строки «mail» (вещественный тип, [0,100])
11.	Частота использования строки «receive» (вещественный тип, [0,100])
12.	Частота использования строки «will» (вещественный тип, [0,100])
13.	Частота использования строки «people» (вещественный тип, [0,100])
14.	Частота использования строки «report» (вещественный тип, [0,100])
15.	Частота использования строки «addresses» (вещественный тип, [0,100])
16.	Частота использования строки «free» (вещественный тип, [0,100])
17.	Частота использования строки «business» (вещественный тип, [0,100])
18.	Частота использования строки «email» (вещественный тип, [0,100])
19.	Частота использования строки «you» (вещественный тип, [0,100])
20.	Частота использования строки «credit» (вещественный тип, [0,100])
21.	Частота использования строки «your» (вещественный тип, [0,100])
22.	Частота использования строки «font» (вещественный тип, [0,100])
23.	Частота использования строки «000» (вещественный тип, [0,100])
24.	Частота использования строки «money» (вещественный тип, [0,100])
25.	Частота использования строки «hp» (вещественный тип, [0,100])
26.	Частота использования строки «hpl» (вещественный тип, [0,100])
27.	Частота использования строки «george» (вещественный тип, [0,100])
28.	Частота использования строки «650» (вещественный тип, [0,100])
29.	Частота использования строки «lab» (вещественный тип, [0,100])
30.	Частота использования строки «labs» (вещественный тип, [0,100])
31.	Частота использования строки «telnet» (вещественный тип, [0,100])
32.	Частота использования строки «857» (вещественный тип, [0,100])
33.	Частота использования строки «data» (вещественный тип, [0,100])
34.	Частота использования строки «415» (вещественный тип, [0,100])
35.	Частота использования строки «85» (вещественный тип, [0,100])
36.	Частота использования строки «technology» (вещественный тип, [0,100])
37.	Частота использования строки «1999» (вещественный тип, [0,100])
38.	Частота использования строки «parts» (вещественный тип, [0,100])
39.	Частота использования строки «pm» (вещественный тип, [0,100])
40.	Частота использования строки «direct» (вещественный тип, [0,100])
41.	Частота использования строки «cs» (вещественный тип, [0,100])
42.	Частота использования строки «meeting» (вещественный тип, [0,100])
43.	Частота использования строки «original» (вещественный тип, [0,100])
44.	Частота использования строки «project» (вещественный тип, [0,100])
45.	Частота использования строки «re» (вещественный тип, [0,100])
46.	Частота использования строки «edu» (вещественный тип, [0,100])
47.	Частота использования строки «table» (вещественный тип, [0,100])
48.	Частота использования строки «conference» (вещественный тип, [0,100])
49.	Частота использования символа “<” (вещественный тип, [0,100])
50.	Частота использования символа “(” (вещественный тип, [0,100])
51.	Частота использования символа “[” (вещественный тип, [0,100])
52.	Частота использования символа “!” (вещественный тип, [0,100])
53.	Частота использования символа “$” (вещественный тип, [0,100])
54.	Частота использования символа “#” (вещественный тип, [0,100])
55.	Средняя длина непрерывной последовательности заглавных букв (вещественный тип, [1, …])
56.	Самая длинная непрерывная последовательность заглавных букв (целый тип, [1, …])
57.	Сумма длин всех непрерывных последовательностей заглавных букв (целый тип, [1, …])

## Классы
0 – не спам, 1 – спам.

# Ход работы

В качестве объектов исследования были взяты два метода машинного обучения: наивный байесовский классификатор (NB) и метод опорных векторов с параметрами, подобранными в предыдущей лабораторной работе (SVM).

In [1]:
# Импорт библиотек
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from time import time

In [2]:
# Загрузка данных из файла
data = pd.read_csv('data/spambase.data', header = None).values
# Разделение на классы и атрибуты
X = data[:, :-1]
Y = np.ravel(data[:, -1:].astype(np.int64, copy=False))
# Выделение обучающей и тестовой выборки
x_train, x_test, y_train, y_test = train_test_split(
    X, Y, test_size=0.3, random_state=42)

## Точность классификации

In [3]:
bayesClf = GaussianNB()
svmClf = SVC(kernel='rbf', C=10, gamma=0.001, probability=True)
bayesAcc = cross_val_score(bayesClf, X, Y, cv=5, scoring='accuracy')
svmAcc = cross_val_score(svmClf, X, Y, cv=5, scoring='accuracy')
print("Доля правильных ответов NB: {:.2f} +/- {:.2f}".format(bayesAcc.mean(), bayesAcc.std()))
print("Доля правильных ответов SVM: {:.2f} +/- {:.2f}".format(svmAcc.mean(), svmAcc.std()))

Доля правильных ответов NB: 0.83 +/- 0.06
Доля правильных ответов SVM: 0.85 +/- 0.05


## Логарифм функции правдоподобия

In [4]:
bayesClf.fit(x_train, y_train)
svmClf.fit(x_train, y_train)
bayesLogLoss = cross_val_score(bayesClf, X, Y, cv=5, scoring='neg_log_loss')
svmLogLoss = cross_val_score(svmClf, X, Y, cv=5, scoring='neg_log_loss')
print("Логарифм функции правдоподобия NB: {:.2f} +/- {:.2f}".format(bayesLogLoss.mean(), bayesLogLoss.std()))
print("Логарифм функции правдоподобия SVM: {:.2f} +/- {:.2f}".format(svmLogLoss.mean(), svmLogLoss.std()))

Логарифм функции правдоподобия NB: -4.44 +/- 1.76
Логарифм функции правдоподобия SVM: -0.36 +/- 0.12


## Область под кривой ошибок

In [5]:
bayesAuc = cross_val_score(bayesClf, X, Y, cv=5, scoring='roc_auc')
svmAuc = cross_val_score(svmClf, X, Y, cv=5, scoring='roc_auc')
print("Область под кривой ошибки NB: {:.2f} +/- {:.2f}".format(bayesAuc.mean(), bayesAuc.std()))
print("Область под кривой ошибки SVM: {:.2f} +/- {:.2f}".format(svmAuc.mean(), svmAuc.std()))

Область под кривой ошибки NB: 0.94 +/- 0.03
Область под кривой ошибки SVM: 0.92 +/- 0.04


## Матрица неточностей

In [6]:
bayesPredict = bayesClf.predict(x_test)
svmPredict = svmClf.predict(x_test)
bayes_matrix = confusion_matrix(y_test, bayesPredict)
svm_matrix = confusion_matrix(y_test, svmPredict)
print("Матрица неточностей NB:\n", bayes_matrix)
print("Матрица неточностей SVM:\n", svm_matrix)

Матрица неточностей NB:
 [[592 212]
 [ 30 547]]
Матрица неточностей SVM:
 [[725  79]
 [100 477]]


## Отчёт классификации

In [7]:
bayes_report = classification_report(y_test, bayesPredict)
svm_report = classification_report(y_test, svmPredict)
print("Отчёт классификации NB:\n", bayes_report)
print("Отчёт классификации SVM:\n", svm_report)

Отчёт классификации NB:
              precision    recall  f1-score   support

          0       0.95      0.74      0.83       804
          1       0.72      0.95      0.82       577

avg / total       0.86      0.82      0.83      1381

Отчёт классификации SVM:
              precision    recall  f1-score   support

          0       0.88      0.90      0.89       804
          1       0.86      0.83      0.84       577

avg / total       0.87      0.87      0.87      1381



# Вывод

Исследование показало, что метод опорных векторов (SVM) превосходит наивный байесовский классификатор (NB) по показателям доли правильных ответов (SVM: 0.85, NB: 0.83), точности (SVM: 0.87, NB: 0.86) и полноты (SVM: 0.87, NB: 0.82). Несмотря на весьма незначительную разницу, метод опорных векторов продемонстрировал лучшие показатели качества, поэтому его использование в данной задаче более предпочтительно, чем использование наивного байесовского классификатора.